eks/ibl_pupil_smoother.py

```diff
@@ -305,133 +308,175 @@ def ensemble_kalman_smoother_ibl_pupil(
-               negative log-likelihoods, negative log-likelihood values.
+        s_finals (list[float]), ms (T,3), Vs (T,3,3), nll (float)
     """
 
-    @partial(jit)
-    def nll_loss_sequential_scan(
-            s_log, ys, m0, S0, C, R, ensemble_vars, diameters_var, x_var, y_var):
-        s = jnp.exp(s_log)  # Ensure positivity
-        return pupil_smooth(
-            s, ys, m0, S0, C, R, ensemble_vars, diameters_var, x_var, y_var)
+    # logistic reparam to keep s in (eps,1-eps)
+    def _to_stable_s(u, eps=1e-3):
+        return jax.nn.sigmoid(u) * (1.0 - 2 * eps) + eps
 
-    loss_function = nll_loss_sequential_scan
-    # Optimize smooth_param
+    # crop ys and ev for the loss if s_frames provided
+    if s_frames and len(s_frames) > 0:
+        y_cropped = crop_frames(ys, s_frames)              # (T', 8)
+        ev_cropped = crop_frames(ensemble_vars, s_frames)  # (T', 8)
+    else:
+        y_cropped, ev_cropped = ys, ensemble_vars
+
+    # build time-varying R_t for loss and for final smoothing (full sequence)
+    R_loss = build_R_from_vars(ev_cropped)     # (T' ,8,8)
+    R_full = build_R_from_vars(ensemble_vars)  # (T  ,8,8)
+
+    # jnp once
+    y_c = jnp.asarray(y_cropped)
+    R_loss = jnp.asarray(R_loss)
+    m0_j, S0_j, C_j = jnp.asarray(m0), jnp.asarray(S0), jnp.asarray(C)
+    y_full = jnp.asarray(ys)
+    R_full = jnp.asarray(R_full)
+
+    # local params builder using your NLGSSM wrapper; pass Q_exact with s=1.0
+    def _params_linear(m0, S0, A, Q_exact, R_any, C):
+        f_fn = (lambda x, A=A: A @ x)
+        h_fn = (lambda x, C=C: C @ x)
+        return params_nlgssm_for_keypoint(m0, S0, Q_exact, 1.0, R_any, f_fn, h_fn)
+
+    # EKF NLL for a given unconstrained u = [u_d, u_c]
+    def _nll_from_u(u: jnp.ndarray) -> jnp.ndarray:
+        s_d, s_c = _to_stable_s(u)
+        A = jnp.diag(jnp.array([s_d, s_c, s_c]))
+        Q = jnp.diag(jnp.array([
+            diameters_var * (1.0 - s_d**2),
+            x_var * (1.0 - s_c**2),
+            y_var * (1.0 - s_c**2),
+        ]))
+        params = _params_linear(m0_j, S0_j, A, Q, R_loss, C_j)
+        post = extended_kalman_filter(params, y_c)
+        return -post.marginal_loglik
+
+    optimizer = optax.adam(lr)
     if smooth_params is None or smooth_params[0] is None or smooth_params[1] is None:
-        # Crop to only contain s_frames for time axis
-        y_cropped = crop_frames(ys, s_frames)
-        ensemble_vars_cropped = crop_frames(ensemble_vars, s_frames)
-
-        # Optimize negative log likelihood
-        s_init = jnp.log(jnp.array([0.99, 0.98]))  # reasonable guess for s_finals
-        optimizer = optax.adam(learning_rate=0.005)
-        opt_state = optimizer.init(s_init)
-
-        def step(s, opt_state):
-            loss, grads = jax.value_and_grad(loss_function)(
-                s, y_cropped, m0, S0, C, R, ensemble_vars_cropped, diameters_var, x_var, y_var
+        # init near your old guess (invert logistic)
```

Comment:
same comment as before; put everything inside this `if` block (including `_params_linear` and `_nll_from_u` above) into a new function called `pupil_optimize_smooth_param` (to mirror the new function in core.py`)